### PR TITLE
[SPARK-33192][CORE]fix Executor will not be deleted when idle timeout while running on k…

### PR DESCRIPTION
**Pull request by hzyangkai**
…8s whith dynamic resource allocation

### What changes were proposed in this pull request?
when the tracker of one executor have no active shuffle ,clear shuffleIds in it.

### Why are the changes needed?
when one job with shuffle operation was completed ,if hasActiveShuffle variable of the tracker of one executor change from true to false,then shuffleIds should be cleared.because shuffleIds is used to decide if the executor has shuffle files.if shuffleIds is not cleared ,in Tractor#updateTimeout function ,newDeadline will be Long.MaxValue,and then executor will not been deleted forever.

### Does this PR introduce _any_ user-facing change?
No

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@hzyangkai, can you file a JIRA and link it to the PR title? See also http://spark.apache.org/contributing.html

**Comment by hzyangkai** [comment]
> @hzyangkai, can you file a JIRA and link it to the PR title? See also http://spark.apache.org/contributing.html

thanks, done

**Comment by hzyangkai** [comment]
test failed not related to this change, what should i do  @HyukjinKwon

**Comment by hzyangkai** [comment]
Jenkins, retest this please

**Comment by hzyangkai** [comment]
hi, @mridulm ,could you help me review this pr please? thanks very much.

**Comment by tgravescs** [comment]
I haven't reviewed yet, but test failure is in executor monitor tests so likely  related to this change, please take a look.

**Comment by hzyangkai** [comment]
> I haven't reviewed yet, but test failure is in executor monitor tests so likely related to this change, please take a look.

thanks for your reply, i have some idea for the test failure.
I think the parameter spark.dynamicAllocation.shuffleTracking.timeout is unreasonable.

According to the spark documentation：this option can be used to control when to time out executors even when they are
storing shuffle data.it indicates that the downstream task will fail if the execution time exceeds this configuration.

But I think this is more reasonable: if all shuffles contained in an executor are not used by an active job, then the executor can be deleted at this time.

i am Looking forward to your reply.

**Comment by tgravescs** [comment]
I don't want to change the definition of that config at this point. I don't see why your change here would be affected by that?  Other then are you saying without this change the executors stay around and is able to shuffle the data for this test?

File: core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala [hunk 1]
```diff
@@ -586,6 +586,7 @@ private[spark] class ExecutorMonitor(
       val hadActiveShuffle = hasActiveShuffle
       hasActiveShuffle = ids.exists(shuffleIds.contains)
       if (hadActiveShuffle && isIdle) {
+        shuffleIds.clear()
```
Comment (tgravescs) [inline core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala:589]: I don't follow this change.  Here isIdle means we don't have any running tasks and hasActiveshuffle is false. The ids passed in are supposed to be all the active ones which is determined by shuffleToActiveJobs.  It does not mean that shuffleIds is empty. shuffleIds stick around until we get the ShuffleCleanedEvent that comes from context cleaner when RDD isn't referenced anymore. If its referenced even though the job may be done it means someone could use that still so we don't remove it until after the timeout.  Here is we are idle and had active shuffle we update the Timeout, which should be setting it to either cache or storage timeout and then when the ExecutorMonitor comes around to check it should see that its set and time it out regardless of if it has shuffle ids.

If there is a bug in there somewhere then  we need to figure out where it is. I do not think is not the correct fix.

**Comment by tgravescs** [comment]
there is not enough information in the jira or description for someone to reproduce, can you please either write a test (ideal if a fix is going to go in) or at least provide manual steps to reproduce.

File: core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala (hunk 1 above)
Comment (hzyangkai) [reply core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala:589]: thank you for your explaining. i got it .it seems works well . i will close this issue.

**Pull request closed without merging**